Merge udacity/warp-pipe changes to upstream

perangel opened the pull request.

Commit:
chore: Add .idea/ to .gitignore

Commit:
feat: Adds the Axon Shutdown method

The Axon Shutdown method has been added for external projects to use.

Commit:
Merge pull request #1 from udacity/bernielomax/feat/axon-shutdown

feat: Axon shutdown

Commit:
refactor: Warp-pipe gets its own database connection

Listeners were the only component that connects to the database. So that global functions
can be added. Warp-Pipe should have its own connection.

Commit:
feat: Axon to auto-shutdown the sync when complete

Commit:
feat: Change events include the changeset id

This is so that Axon can auto-shutdown by checking the change event id matches that of the last record in the changeset table.

Commit:
Merge pull request #2 from udacity/bernielomax/feat/change-event-includes-id

feat: Change events include the changeset id

Commit:
Merge pull request #3 from udacity/bernielomax/refactor/warp-pipe-db-connection

refactor: Warp-Pipe db connection

Commit:
Merge pull request #4 from udacity/bernielomax/feat/axon-shutdown-on-sync

feat: Axon shutdown on sync

Commit:
feat: Only create even triggers when they do not exist

Makes db.Prepare repeatable by preventing the following error:

  ERROR: trigger "public__foo_changesets" for relation "foo" already exists (SQLSTATE 42710)+

Commit:
Merge pull request #5 from udacity/bernielomax/feat/create-trigger-when-not-exists

feat: Create trigger when not exists

Commit:
refactor: axon.Run returns rather than os.Exit

To avoid exiting apps that use axon.Run unnecessarily. Return rather than os.Exit.

svysali committed:
Merge pull request #6 from udacity/bernielomax/refactor/do-not-os-exit

refactor: Axon run does not os.Exit

Commit:
refactor: Export table list generation

I would like to be able to use the same logic in other external projects.

Commit:
Merge pull request #8 from udacity/bernielomax/refactor/export-table-list-generation

refactor: Export table list generation

Commit:
fix: mixed-case postgres table names require quotes.

Postgres does not recognize mixed case table/database names and
automatically converts them to lower case. This results in a
relation/database not found error.
Wrap database and table names in double quotes, so that they
are parsed correctly.

Commit:
refactor: GenerateTablesList now also returns primary key name and fields

perangel commented on axon.go:
```diff
@@ -98,46 +107,68 @@ func (a *Axon) Run() {
 		case change := <-changes:
 			a.processChange(sourceDBConn, targetDBConn, a.Config.TargetDBSchema, change)
-
+			isLatest, err := wp.IsLatestChangeSet(change.ID)
```
This is a useful feature, but I don't think it should be the default behavior. If you wanted to run a long-lived axon process to keep replicating changes to a downstream DB, then this would halt the process the minute the target database was caught up.

bernielomax replied:
Agreed. I will look to make this configurable.

perangel commented on db/schema.go:
```diff
@@ -155,48 +156,122 @@ func createTriggerFunc(tx *pgx.Tx) error {
+	}
+	table.PKeyFields = make(map[int]string)
+	rows, err := conn.Query(`
+		SELECT 
```
We uppercase all SQL keywords in the project, so I would please ask that you uppercase all keywords (i.e. FROM, JOIN, AS, SELECT, WHERE, etc). This helps with readability.

perangel commented on warp_pipe.go:
```diff
@@ -163,6 +164,25 @@ func (w *WarpPipe) Close() error {
 }
 
+// IsLatestChangeSet returns true if the id argument matches that of the last record in the changeset table.
+func (w *WarpPipe) IsLatestChangeSet(id int64) (bool, error) {
```
This implementation will only work for "audit" mode (where we are using `LISTEN/NOTIFY`) and will not work for logical replication mode. So we should perhaps make WP panic if you are running logical replication mode and have enabled the flag for auto-shutdown-on-complete-sync behavior. (see: commend above)

bernielomax replied:
Yes. This is something I need to fix up. I think it needs to be moved to the listener interface? Was going to return a not supported error for LR? Unless you know of a way to determine this?

Commit:
refactor: Make Axon exiting on latest optional

Commit:
feat: add checksum based verification to axon

Commit:
fix: JSON fails checksum

Maps cannot be re-marshaled. It can cause differences between the source and target JSON. To avoid this
from happening maps are no longer a supported changeset column value. Any JSON must be passed as the
original string value. This bug was only affecting the notify listener. The logical replication listener
was already passing JSON as a string.

Commit:
Merge pull request #10 from udacity/bernielomax/refactor/exit-on-latest-optional

refactor: Exit on latest optional

Commit:
Merge pull request #13 from udacity/bernielomax/fix/json-fails-checksum

fix: JSON fails checksum

Commit:
fix: Double quote trigger name when creating

Commit:
improve: use rolling hash aggregation for checksum verification

Previously we collected the md5 hash for each row and ran a md5 over
the collection of hashes. This was non-performant, in the order of minutes
for very large table. This aggregation does a rolling md5 hash, which greatly
improves performance.

source: https://stackoverflow.com/questions/4020033/how-can-i-get-a-hash-of-an-entire-table-in-postgresql

Commit:
Merge pull request #14 from udacity/bernielomax/fix/quote-trigger-name-on-creation

fix: Double quote trigger name when creating

13rac1 commented:
FWIW A `git rebase` will drop of these unneeded `Merge pull request` commits if you want

13rac1 commented on axon.go:
```diff
@@ -98,44 +109,155 @@ func (a *Axon) Run() {
+				if err != nil {
+					a.Logger.WithError(err).
+						WithField("component", "warp_pipe").
+						Fatal("failed to determine if the sync is complete")
+				}
```
Should this return or break?

perangel replied:
`Fatal()` calls `Exit(1)`

13rac1 commented on db/schema.go:
```diff
@@ -155,48 +156,190 @@ func createTriggerFunc(tx *pgx.Tx) error {
+
+	tables := make([]Table, 0)
+	for tableName, include := range tableRegister {
+		if include {
```
nit: an early return (or break in this case) would reduce nesting and increase clarity

13rac1 commented on db/schema.go:
```diff
@@ -155,48 +156,190 @@ func createTriggerFunc(tx *pgx.Tx) error {
+		table.Name = nameArr[0]
+	}
+	table.PKeyFields = make(map[int]string)
+	rows, err := conn.Query(`
```
This error is not handled.

13rac1 commented on db/schema.go:
```diff
@@ -155,48 +156,190 @@ func createTriggerFunc(tx *pgx.Tx) error {
+	sql := fmt.Sprintf(`
+		DO  
+		$$  
+		BEGIN  
+			IF NOT EXISTS(
+                 SELECT * FROM(
+					 SELECT trigger_name AS name, concat_ws('.', event_object_schema, event_object_table) AS table 
+					 FROM information_schema.triggers
+                 ) AS triggers
+				 WHERE triggers.name = '%s'
+				 AND triggers.table = '%s.%s'  
+			)  
+			THEN
+				CREATE TRIGGER "%s"
+				AFTER INSERT OR UPDATE OR DELETE
+				ON "%s"."%s"
+				FOR EACH ROW EXECUTE PROCEDURE warp_pipe.on_modify();
+			END IF ;
+		END;  
```
Tab alignment on this query should be more consistent

Commit:
fix: Prettify changeset JSON

Needed to checksum and validate the source and target changeset tables.

Commit:
Merge pull request #16 from udacity/bernielomax/fix/pretty-changeset-json

fix: Pretty changeset json

Commit:
improve: Setup fails if a table does not have primary key.

perangel committed:
refactor: cleanup for merging to upstream

perangel committed:
Merge pull request #18 from udacity/perangel/refactor/fixes-for-merging-to-upstream

refactor: cleanup for merging to upstream
